### DISCUSSION TITLE
Questions

### DISCUSSION THREAD

**Discussion by dAnjou** [ScalaConsultants/mesmer#273] [Q&A]
Hi 👋 

sorry about the shitty title, I just have some questions regarding the architecture but on different levels, happy to separate them in any way if that's better for you.

What's the plan/motivation/bigger idea behind the separation into a JVM agent and an Akka extension?
> - Akka extension - that runs in the background and is responsible for exporting the metrics to your chosen backend
> - JVM agent - that instruments Akka classes to expose metrics for the extension

The [OpenTelemetry Instrumentation for Java](https://github.com/open-telemetry/opentelemetry-java-instrumentation) provides a JVM agent that does both, it hooks into Akka _and_ you can [expose exporters like Prometheus](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#prometheus-exporter). The setup was trivial, we're currently running it in our system but only for the tracing for now, haven't looked into the metrics yet.

It seems Mesmer provides more metrics which would be nice of course, and especially this makes me think:

> As Mesmer uses OpenTelemetry underneath [...]

Theoretically, could the core of Mesmer be wrapped into an [extension for their agent](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/README.md) 🤔 

Besides that, I'm a bit confused by this whole paragraph because I don't understand Mesmer's intended architecture ([the provided overview](https://github.com/ScalaConsultants/mesmer-akka-agent/blob/0b9207fb27932b483c6dba694679a66a1c1a29c6/extension_overview.png) isn't detailed enough):

> As Mesmer uses OpenTelemetry underneath to export data to metric backend you need to set up an exporter. All exporters require OpenTelemetry SDK present, so make sure you have one added to your project - without this all measurement operations will be NoOp. You can for example set up your project with OTLP metrics exporter that also includes the SDK

While writing this text I'm now understanding more and more but it's quite hard to see where Mesmer fits in, a detailed architecture or data flow diagram would really help.

Hope you can help out 🙏

**Comment by lgajowy** [comment]
> While writing this text I'm now understanding more and more but it's quite hard to see where Mesmer fits in, a detailed architecture or data flow diagram would really help. 

First of all thanks for reaching out! :) Mesmer will be developed and maintained more actively now so it's perfect for us to receive your feedback! 

> Besides that, I'm a bit confused by this whole paragraph because I don't understand Mesmer's intended architecture (the provided overview isn't detailed enough): 

There's a blogpost that may shed some more light when it comes to the architecture: https://scalac.io/blog/akka-monitoring-with-mesmer-and-opentelemetry/

When it comes to the paragraph you quoted - I have to take a deeper look. I will and will try to get back to you soon. 

> The OpenTelemetry Instrumentation for Java provides a JVM agent that does both, it hooks into Akka and you can expose exporters like Prometheus. The setup was trivial, we're currently running it in our system but only for the tracing for now, haven't looked into the metrics yet.

I'm currently evaluating ways of joining the functionality we have in mesmer with the opentelemetry agent. I totally agree that utilizing it would be benefitial for mesmer. Not only it will bring more features but also reduce our codebase. Using extensions is something I will evalutate too (stay tuned).

**Comment by lgajowy** [comment]
> Besides that, I'm a bit confused by this whole paragraph because I don't understand Mesmer's intended architecture ([the provided overview](https://github.com/ScalaConsultants/mesmer-akka-agent/blob/0b9207fb27932b483c6dba694679a66a1c1a29c6/extension_overview.png) isn't detailed enough):
> 
> > As Mesmer uses OpenTelemetry underneath to export data to metric backend you need to set up an exporter. All exporters require OpenTelemetry SDK present, so make sure you have one added to your project - without this all measurement operations will be NoOp. You can for example set up your project with OTLP metrics exporter that also includes the SDK

In the meantime: @worekleszczy do you think you could help here? :)

**Comment by lgajowy** [comment]
> Theoretically, could the core of Mesmer be wrapped into an extension for their agent 🤔

> I'm currently evaluating ways of joining the functionality we have in mesmer with the opentelemetry agent.

@dAnjou I checked and the direction to go with the extension instead of a custom agent seems to be a good one imho. Thanks for bringing it up as well. Please see: https://github.com/ScalaConsultants/mesmer-akka-agent/discussions/272 if you're curious/have further comments.


> Besides that, I'm a bit confused by this whole paragraph because I don't understand Mesmer's intended architecture (the provided overview isn't detailed enough):

If you don't add the exporter with the sdk, mesmer will use noop instruments so you will get nothing. This is because mesmer uses opentelemetry-api only. Providing the exporter is currently user's responsibility. 
I fixed the docs a little bit - feel free to take a look: https://github.com/ScalaConsultants/mesmer-akka-agent#exporter

**Reply by dAnjou** [reply]
Hey, thanks for following up!

We've been evaluating a few options but with other things on our plate things are a bit slow, sorry.
